### PR TITLE
FW-1221: Support Arrays for LhsValue

### PR DESCRIPTION
**Pull request by grobinson-cloudflare**

### CODE REVIEW

File: engine/src/rhs_types/array.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+use crate::{
+    lex::{Lex, LexResult},
+    strict_partial_ord::StrictPartialOrd,
+    types::{Array, GetType, Type},
+};
+use serde::Serialize;
+use std::{borrow::Borrow, cmp::Ordering};
+
+/// [Uninhabited / empty type](https://doc.rust-lang.org/nomicon/exotic-sizes.html#empty-types)
+/// for `map` with traits we need for RHS values.
```
Comment (marmeladema) [inline engine/src/rhs_types/array.rs]: You meant for `array`?

File: engine/src/types.rs [hunk 2]
```diff
@@ -196,6 +199,64 @@ macro_rules! declare_types {
     };
 }
 
+/// type Array<'a> = Vec<LhsValue<'a>>;
```
Comment (marmeladema) [inline engine/src/types.rs]: Maybe a comment for similar to the one for `Map`? Something like ` /// An array of [`Type`].`

File: engine/src/types.rs [hunk 3]
```diff
@@ -196,6 +199,64 @@ macro_rules! declare_types {
...
+    ) -> Result<Option<LhsValue<'a>>, TypeMismatchError> {
+        let value_type = value.get_type();
+        if self.val_type != value_type {
+            return Err(TypeMismatchError {
+                expected: self.val_type.clone(),
+                actual: value_type,
+            });
+        }
+        self.data.insert(idx, value.clone());
+        Ok(Some(value))
```
[the remark this answers is not in the document]
Comment (marmeladema) [reply engine/src/types.rs]: It cannot really be coherent with `Map` behavior that returns the previous value stored at this index. I think we should not return previous value and change return type of `insert` functions to be something like `Result<(), TypeMismatchError>`

File: engine/src/types.rs [hunk 4]
```diff
@@ -318,6 +386,13 @@ impl<'a> LhsValue<'a> {
         let value = value.into();
         let value_type = value.get_type();
         match item {
+            FieldIndex::ArrayIndex(idx) => match self {
+                LhsValue::Array(ref mut arr) => arr.insert(idx as usize, value),
+                _ => Err(TypeMismatchError {
+                    expected: Type::Map(Box::new(value_type)),
```
Comment (marmeladema) [inline engine/src/types.rs]: I think you are expecting a `Type::Array`

File: engine/src/types.rs [hunk 5]
```diff
@@ -298,8 +361,13 @@ impl<'a> LhsValue<'a> {
     /// nested element. Only LhsValue::Map supports nested elements for now.
```
Comment (marmeladema) [inline engine/src/types.rs]: Maybe update the comments now that `Array` also supports nested elements?

File: engine/src/scheme.rs [hunk 6]
```diff
@@ -18,12 +19,50 @@ use std::{
...
+/// FieldIndex is an enum with variants [`ArrayIndex(usize)`],
+/// representing an index into an Array, or `[MapKey(String)`],
+/// representing a key into a Map.
+///
+/// ```
+/// #[allow(dead_code)]
+/// enum FieldIndex {
+///     ArrayIndex(u32),
+///     MapKey(String),
+/// }
```
[the remark this answers is not in the document]
Comment (grobinson-cloudflare) [reply engine/src/scheme.rs]: I was following https://doc.rust-lang.org/src/core/result.rs.html#1-15

[the remark this answers is not in the document]
Comment (marmeladema) [reply engine/src/scheme.rs]: Ok then

**Pull request merged**